numpy/core/src/multiarray/textreading/str_to_int.c

```diff
@@ -41,8 +63,24 @@
         uint64_t parsed;                                                    \
         uintw##_t x;                                                        \
                                                                             \
-        if (str_to_uint64(str, end, UINT_MAX, &parsed) < 0) {               \
-            return -1;                                                      \
+        if (NPY_UNLIKELY(                                                   \
+                str_to_uint64(str, end, UINT_MAX, &parsed) < 0)) {          \
+            /* DEPRECATED 2022-07-03, NumPy 1.23 */                         \
+            double fval;                                                    \
+            PyArray_Descr *d_descr = PyArray_DescrFromType(NPY_DOUBLE);     \
+            Py_DECREF(d_descr);  /* borrowed */                             \
+            if (to_double(d_descr, str, end, (char *)&fval, pconfig) < 0) { \
+                return -1;                                                  \
+            }                                                               \
+            if (!pconfig->gave_int_via_float_warning) {                     \
+                pconfig->gave_int_via_float_warning = true;                 \
+                if (PyErr_WarnEx(PyExc_DeprecationWarning,                  \
+                        deprecation_msg, 3) < 0) {                          \
+                    return -1;                                              \
+                }                                                           \
+            }                                                               \
+            pconfig->gave_int_via_float_warning = true;                     \
+            x = (uintw##_t)fval;                                            \
```

Comment:
Now that we have C++, this should even more be cleaned up and stop being a macro.  But I don't think it is important enough to try that now.  Especially on something that needs backporting.